### ISSUE TITLE
[GOV] security architecture for release and community assets - separation of duties, MFA 

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#3836]
`sktime` is starting to attract a sizeable community on its social channels (many 100s) and is starting to have a substantial reverse dependency footprint.

We should therefore consider safeguards against malicious actors in our operations model, both from external and internal.

Two critical points in `sktime`'s security architecture are imo currently:

* the release process, and control of the `pypi` wheels, e.g., for a [software supply chain attack](https://en.wikipedia.org/wiki/Supply_chain_attack). These have become more frequent recently.
* community assets, communication and operations channels including GitHub accounts of developers, control of the repositories, and social channels Slack, Discord, twitter

Motives for attackes can differ, e.g.:

* criminal actors using `sktime` as a stepping stone for ransomware attacks via SSCA
* nation state actors using `sktime` for SSCA, e.g., in the context of cyberattacks
* community assets used as leverage in community internal social conflicts 
* community assets being seized as a supportive measure to enhance length, scope, or scale of an SSCA

In my opinion, the key tools to mitigate any of the above risks are:

* [separation of duties](https://en.wikipedia.org/wiki/Separation_of_duties) - reducing the risk that a single actor can seize control of an entire process of all assets, by increasing the number of separate responsibilities or control points
* [multi-factor authentication](https://en.wikipedia.org/wiki/Multi-factor_authentication) - reducing the risk of impersonation of elevated rights by a malicious actor by increasing the number of independent authentication points

This issue is also about discussing:

* are there other key critical points that need to be covered by security architecture?
* are there other measures than the above that we should consider?
* what are the trade-offs, e.g., with operational efficiency
* what is the best implementation of said measures?

**Comment by fkiraly** [comment]
Mapping community assets and maximal roles, with worst case action radius:

* GitHub organisation admin account. This role "trumps" GitHub repo admins. It can seize control of all GitHub assets by removing other admins. There can be multiple GitHub repo or org admins.
* pypi account. This role can seize control of the package on pypi. There can be only one such account afaik.
* Slack primary owner. This role can seize control of the entire slack space. There can be only one such role holder.
* Discord primary owner. This role can seize control of the entire discord space. There can be only one such role holder.
* personal email accounts associated with the above roles. 
* sktime email address accounts.
* not sure about linkedin, twitter

Regarding trump relations:
* control of certain email accounts, under certain conditions, allow seizing control of some of the above accounts. If the same email account is used for multiple primary accounts, it may allow seizing of multiple assets and can be a weak point.
* GitHub maximal role does not allow seizure of pypi (secrets cannot be read afaik), but it allows pushing a package release.

**Comment by fkiraly** [comment]
Mapping release process, we have relevant roles and process element:

* change code on `main`. Anyone with write access can merge to `main` if CI passes and an approval is given. Anyone with admin access can merge rigiht away.
* change requirements in CI for release on pypi: same conditions, because it's infrastructure as code in the same repo (GHA)
* trigger a release: the release manager, formally. But practially, anyone with `sktime` repository write access can trigger a release to `pypi`. That is all the (currently 17) `sktime` core developers.
* GitHub release notifications on slack general channel: can be deactivated or deleted by slack admins
* halt a release: same credentials as trigger a release
* control of pypi release endpoint: pypi account owner, that role can replace the current `sktime` release with arbitrary code

From a security architecture perspective, ironically, tests that take longer come with a higher degree of safety, because a longer duration causes a delay between process element start and end which increases likelihood of detection.

The weak points of the release process seem to be:

* core devs and admin roles on the GitHub repo
* owner of the `sktime` pypi account
* notification of process start and end (or absence), slack admins
* duration of GitHub release, i.e., between release start and release endpoint (upload to pypi)

**Comment by fkiraly** [comment]
the most promising strategies for an attacker seem to be:

* seize control of `pypi` account, replace the wheels. Can even be done intermittently in sync with cron jobs of a targeted entity, i.e., replace with malicious ones and then replace back again with the correct ones.
* seize control of GitHub admin role plus slack admin role. Push to `main` a change which: removes CI checks to reduce test time from 1-2h to nil; any malicious code. Trigger a pypi upload, delete any release notifications on GitHub and slack.

**Comment by fkiraly** [comment]
Possible actions to make us more secure.
Any thoughts on this or other measures?

Quick wins:

* enforce 2FA for all developers, this reduces risk of impersonation - multi-factor principle
* ensure no single email address or person holds multiple of the maximal-admin roles (GitHub, pypi, discord, slack, sktime emails, social media) - separation of duties
* diversify locations of release process start notifications, e.g., slack and discord both - separation of duties

More complex:

* enforce IP ranges for GitHub organisation, based on known/trusted locations of core developers
* introduce time delays for release process elements, i.e., between notification and endpoint

**Comment by fkiraly** [comment]
Question: is there any way to introduce separation of duties principle for spaces that have only one super-admin, e.g., slack, discord?

**Comment by fkiraly** [comment]
Question: is there anything we can do about the `pypi` account which seems like a single-vector, no-delay, no-notification point? Surely someone must have thought about this, there are far more important python packages than `sktime` that live on pypi...

**Comment by jernsting** [comment]
> pypi account. This role can seize control of the package on pypi. There can be only one such account afaik.

There can be multiple accounts for a package. See for example https://pypi.org/project/photonai/ (Maintainers). But all maintainers of a package have the same role and can add or remove existing package maintainers.

**Comment by jernsting** [comment]
An an additional potential countermeasure would be reducing the used services. For example readthedocs __could__ be replaced by github pages.

**Comment by fkiraly** [comment]
> But all maintainers of a package have the same role and can add or remove existing package maintainers.

Yes, that's just as bad as having a single account in terms of security risk - in fact it has a *higher* risk, since multiple accounts have a higher risk of becoming compromised.

>An an additional potential countermeasure would be reducing the used services.

It may seem counterintuitive, but a more homogenous service stack has a higher risk of becoming compromised all at once.

E.g., if it is all GitHub, any GitHub becomes a potential entry point for takeover of code base *and* documentation.
More precisely, an attacker would have to invest more work to take over both if they live on different services, so it increases the cost for the attacker.

It seems to be counterintuitive since it also increases cost for a benevolent maintainer, but the security perspective is for the "opposite" persona, i.e., malevolent attacker.
Many security measures increase cost for both, and a good security measure has a lot of increase for a malicious attacker per unit of increase for maintainers, users, etc - an example would be 2FA, which increases maintainer/user discomfort only a little but, but doubles the effort for the attacker.

**Comment by jernsting** [comment]
> It may seem counterintuitive, but a more homogenous service stack has a higher risk of becoming compromised all at once.

1. Question is, if the external services are better secured than github. Using multiple secure E-Mail Accounts with secure passwords for different services is in my opinion most commonly less secure than using secure credentials for one service.

2. If one is able to take over the github account, would the documentation on readthedocs really be safe? Wouldn't an intruder be able to push a compromised version of the documentation? If so, the risk of becoming compromised is not reduced by using multiple services.

**Comment by fkiraly** [comment]
Some updates arising from recent learnings about pypi accounts:

- 2FA is automatically enabled for "critical" projects (such as sktime), critical is defined as top 1% percentile of download count among pypi packages
- it is *not* possible to replace wheels of the same file name. However, it may be possible to replace wheels with the same version metadata but a different filename (tbc)
- pypi accounts support the n-key protocol (n people required for authorization) in two ways: TOTP compatible apps, which in-principle can support it as a software feature; use of a USB authentication device which can be subjected to the protocol in the physical world (e.g., putting it into an n-key safe or a bank vault) - see here for pypi authentication: https://pypi.org/help/#totp

**Comment by fkiraly** [comment]
Based on the above, we could do:

* use API token for the github-to-sktime release pipeline
* for the pypi acct, the most secure would be: (i) use USB device, (ii) put it into a real bank vault, (iii) write a contract who can access it, e.g., requires majority of council as per adopted statutes.
* perhaps some provisions on short term decision making, there's clearly a risk trade-off there between reaction time and risk of the acct getting compromised
